packages/engine/Source/Core/TerrainPicker.js

```diff
@@ -0,0 +1,610 @@
+    const v2 = getVertexPosition(
+      encoding,
+      mode,
+      projection,
+      vertices,
+      indices[3 * triIndex + 2],
+      scratchTrianglePoints[2],
+    );
+
+    const triT = IntersectionTests.rayTriangleParametric(
+      ray,
+      v0,
+      v1,
+      v2,
+      cullBackFaces,
+    );
+
+    if (defined(triT) && triT < result && triT >= 0) {
+      result = triT;
+    }
+
+    if (shouldBuildChildren) {
+      packTriangleBuffers(
+        trianglePositions,
+        triangleIndices,
+        scratchTrianglePoints,
+        triIndex,
+        i,
+      );
+    }
+  }
+
+  if (shouldBuildChildren) {
+    for (let childIdx = 0; childIdx < 4; childIdx++) {
+      node.addChild(childIdx);
+    }
+
+    addTrianglesToChildrenNodes(
+      terrainPicker._inverseTransform,
+      node,
+      triangleIndices,
+      trianglePositions,
+    );
+  }
+
+  return result;
+}
+
+const scratchCartographic = new Cartographic();
+
+/**
+ * Gets a vertex position from the buffer, taking into account the exaggeration and scene mode of the terrain.
+ *
+ * @param {TerrainEncoding} encoding The terrain encoding.
+ * @param {SceneMode} mode The scene mode (2D/3D/Columbus View).
+ * @param {MapProjection} projection The map projection.
+ * @param {Float32Array} vertices The vertex buffer of the terrain mesh.
+ * @param {Number} index The index of the vertex to get.
+ * @param {Cartesian3} result The decoded, exaggerated, and possibly projected vertex position.
+ * @returns {Cartesian3} The result vertex position.
+ * @private
+ */
+function getVertexPosition(
+  encoding,
+  mode,
+  projection,
+  vertices,
+  index,
+  result,
+) {
+  let position = encoding.getExaggeratedPosition(vertices, index, result);
+  if (mode === SceneMode.SCENE3D) {
+    return position;
+  }
+
+  const ellipsoid = projection.ellipsoid;
+  const positionCartographic = ellipsoid.cartesianToCartographic(
+    position,
+    scratchCartographic,
+  );
+  position = projection.project(positionCartographic, result);
+  // Swizzle because coordinate basis are different in 2D/Columbus View
+  position = Cartesian3.fromElements(
+    position.z,
+    position.x,
+    position.y,
+    result,
+  );
+
+  return position;
+}
+
+/**
+ * Adds triangles to the child nodes of the given node by launching a worker process to do AABB-triangle testing.
+ *
+ * @param {Matrix4} inverseTransform
+ * @param {TerrainNode} node
+ * @param {Uint32Array} triangleIndices
+ * @param {Float32Array} trianglePositions
+ * @returns {Promise<void>} A promise that resolves when the triangles have been added to the child nodes.
+ * @private
+ */
+async function addTrianglesToChildrenNodes(
+  inverseTransform,
+  node,
+  triangleIndices,
+  trianglePositions,
+) {
+  node.buildingChildren = true;
+
+  // Prepare data to be sent to a worker
+  const matrixArray = new Array(16);
+  Matrix4.pack(inverseTransform, matrixArray, 0);
+  const inverseTransformPacked = new Float64Array(matrixArray);
```

Comment:
I think this can be simplified to:

```suggestion
  // Prepare data to be sent to a worker
  const inverseTransformPacked = new Float64Array(16);
  Matrix4.pack(inverseTransform, inverseTransformPacked, 0);
```